KAFKA-18681 Created GetReplicaLogInfo* RPCs

josefk31 opened the pull request:
Creates GetReplicaLogInfoRequest and GetReplicaLogInfoResponse RPCs
Information returned by these brokers will be used to aid
unclean-recovery by selecting longest logs.

RPCs are cluster actions, meaning they may only be queried by
controllers to other brokers.

Delete this text and replace it with a detailed description of your
change. The   PR title and body will become the squashed commit message.

See
[KIP-966](https://cwiki.apache.org/confluence/display/KAFKA/KIP-966%3A+Eligible+Leader+Replicas#KIP966:EligibleLeaderReplicas-GetReplicaLogInfoRequest(ComingwithUncleanRecovery))
for more information on this RPC.

josefk31 committed:
[KAFKA-18681] Created GetReplicaLogInfo* RPCs

Creates GetReplicaLogInfoRequest and GetReplicaLogInfoResponse RPCs
Information returned by these brokers will be used to aid unclean-recovery by selecting longest logs.

RPCs are cluster actions, meaning they may only be queried by controllers to other brokers.

TaiJuWu commented on clients/src/main/resources/common/message/GetReplicaLogInfoRequest.json:
```diff
@@ -0,0 +1,34 @@
+        { "name": "Partitions", "type": "[]int32", "versions": "0+",
+          "about": "The partitions of this topic whose leader should be elected." }
+      ]}
+  ] }
```
Could you add new line?

TaiJuWu commented on clients/src/main/resources/common/message/GetReplicaLogInfoResponse.json:
```diff
@@ -0,0 +1,39 @@
+            { "name": "ErrorCode", "type": "int16", "versions": "0+", "about": "The result error, or zero if there was no error."},
+            { "name": "ErrorMessage", "type": "string", "versions": "0+", "nullableVersions": "0+", "about": "The result message, or null if there was no error."}      ]}
+      ]}
+  ] }
```
ditto

CalvinLiu7947 commented on clients/src/main/java/org/apache/kafka/common/requests/GetReplicaLogInfoRequest.java:
```diff
@@ -0,0 +1,98 @@
+            this.data = data;
+        }
+
+        public Builder(List<GetReplicaLogInfoRequestData.TopicPartitions> topicPartitions) {
```
Do we need a method to set the broker ID?

josefk31 replied:
Good point; I removed it.

CalvinLiu7947 commented on clients/src/main/resources/common/message/GetReplicaLogInfoRequest.json:
```diff
@@ -0,0 +1,34 @@
+  "validVersions": "0",
+  "flexibleVersions": "0+",
+  "fields": [
+    { "name": "BrokerId", "type": "int32", "versions": "0+", "entityType": "brokerId",
```
I wonder if this broker ID is really necessary. It seems it is handled on the network layer that only the broker with the ID can receive this request.

CalvinLiu7947 reviewed (commented):
Thanks for the change! Got one question for the request.json

josefk31 committed:
Removed brokerId field since its superflous

josefk31 committed:
Fixed compilation error

cmccabe commented on clients/src/main/resources/common/message/GetReplicaLogInfoRequest.json:
```diff
@@ -0,0 +1,32 @@
+    { "name": "TopicPartitions", "type": "[]TopicPartitions", "versions": "0+", "nullableVersions": "0+",
+      "about": "The topic partitions to query the log info.",
+      "fields": [
+        { "name": "TopicId", "type": "uuid", "versions": "0+", "ignorable": true, "about": "The unique topic ID"},
```
what would it mean to ignore TopicId?

josefk31 replied:
Good catch; that was there by error. I don't think that should ever be ignorable.

ahuang98 replied:
just a reminder this is still set to ignorable=true

cmccabe commented:
We need a boolean to indicate that there is more data that we were unable to send due to length restrictions

cmccabe commented on clients/src/main/resources/common/message/GetReplicaLogInfoResponse.json:
```diff
@@ -0,0 +1,40 @@
+            { "name": "CurrentLeaderEpoch", "type": "int32", "versions": "0+", "about": "The current leader epoch for the partition from the broker point of view." },
+            { "name": "LogEndOffset", "type": "int64", "versions": "0+", "about": "The log end offset for the partition." },
+            { "name": "ErrorCode", "type": "int16", "versions": "0+", "about": "The result error, or zero if there was no error."},
+            { "name": "ErrorMessage", "type": "string", "versions": "0+", "nullableVersions": "0+", "about": "The result message, or null if there was no error."}      ]}
```
could remove this

cmccabe commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -108,7 +109,8 @@ class KafkaApis(val requestChannel: RequestChannel,
                 val clientMetricsManager: ClientMetricsManager,
-                val groupConfigManager: GroupConfigManager
+                val groupConfigManager: GroupConfigManager,
+                val brokerEpochSupplier: Supplier[java.lang.Long]
```
let's avoid changing this and just get it from the cache

josefk31 committed:
Responded to PR feedback

For Pagination/ Truncation:

1. Added Truncate boolean flag to indicate that we did not return all the topic partitions
2. Refactored the handler method to make this code easier to write

Also removed the "ErrorMessage" flag since we are not actually using it anymore.

CalvinLiu7947 commented:
> We need a boolean to indicate that there is more data that we were unable to send due to length restrictions

I wonder if it is good enough to indicate the limit is reached when the partition is missing from the response? Especially, we use an explicit partition list in the request. @cmccabe

ahuang98 commented on clients/src/main/resources/common/message/GetReplicaLogInfoRequest.json:
```diff
@@ -0,0 +1,32 @@
+  "flexibleVersions": "0+",
+  "fields": [
+    { "name": "TopicPartitions", "type": "[]TopicPartitions", "versions": "0+", "nullableVersions": "0+",
+      "about": "The topic partitions to query the log info.",
```
nit, "to query the log info for."

ahuang98 commented on clients/src/main/java/org/apache/kafka/common/requests/GetReplicaLogInfoResponse.java:
```diff
@@ -0,0 +1,75 @@
+    public static GetReplicaLogInfoResponse parse(Readable readable, short version) {
+        return new GetReplicaLogInfoResponse(new GetReplicaLogInfoResponseData(readable, version));
+    }
+}
```
nit, new line at end of files

ahuang98 commented on clients/src/main/resources/common/message/GetReplicaLogInfoResponse.json:
```diff
@@ -0,0 +1,43 @@
+    { "name": "ThrottleTimeMs", "type": "int32", "versions": "0+",
+      "about": "The duration in milliseconds for which the request was throttled due to a quota violation, or zero if the request did not violate any quota." },
+    { "name": "BrokerEpoch", "type": "int64", "versions": "0+",
+      "about": "The epoch for the broker." },
```
nit: "of" instead of "for"

ahuang98 commented on clients/src/main/resources/common/message/GetReplicaLogInfoResponse.json:
```diff
@@ -0,0 +1,43 @@
+    { "name": "BrokerEpoch", "type": "int64", "versions": "0+",
+      "about": "The epoch for the broker." },
+    { "name": "Truncated", "type": "bool", "versions": "0+",
+      "about": "True if topic partitions were truncated. Only the first 1000 topic partitions are returned."},
```
hm, what about 
"True if response does not include all the topic partitions requested. Only the first 1000 ..."
or maybe
"True if response is missing topic partitions. Only the first 1000 ..."

ahuang98 commented on clients/src/main/resources/common/message/GetReplicaLogInfoResponse.json:
```diff
@@ -0,0 +1,43 @@
+    { "name": "Truncated", "type": "bool", "versions": "0+",
+      "about": "True if topic partitions were truncated. Only the first 1000 topic partitions are returned."},
+    { "name": "TopicPartitionLogInfoList", "type": "[]TopicPartitionLogInfo", "versions": "0+",
+      "about": "The list of the log info.",
```
nit: List of partition log info

ahuang98 reviewed (commented):
partial review, still need to take a look at your tests!

josefk31 committed:
Addressed more PR comments

josefk31 commented:
> > We need a boolean to indicate that there is more data that we were unable to send due to length restrictions
> 
> I wonder if it is good enough to indicate the limit is reached when the partition is missing from the response? Especially, we use an explicit partition list in the request. @cmccabe

@CalvinConfluent I think we're dealing with the following tradeoffs:

1. We want to save space on RPCs
2. We want to reduce processing time on the brokers
3. RPC should be predictable

It's more obvious what the RPC is doing if we return an error code for each partition above the limit. That being said; I think it does not address (1) completely. I think that the `Truncate` is balancing (1) and (2) at the expense of (3) being a bit less clear. It's still fairly straightforward that we return the first 1k requested (in order) and set `Truncate` to true and requires a consumer of the API to read the API definition. But since this is a cluster-action I feel that the confusion is fair to impose?

cmccabe commented on clients/src/main/resources/common/message/GetReplicaLogInfoResponse.json:
```diff
@@ -0,0 +1,43 @@
+  "name": "GetReplicaLogInfoResponse",
+  "validVersions": "0",
+  "flexibleVersions": "0+",
+  "fields": [
```
as per our earlier conversation, you need to include a "has more data" boolean in this response.

josefk31 replied:
Got it; added a `Truncated` field but changed this to `HasMoreData` since that name seems more clear.

josefk31 committed:
PR feedback comments

Changed Truncate => HasMoreData
Removed ignorable from topicId

ahuang98 commented on core/src/test/scala/unit/kafka/server/RequestQuotaTest.scala:
```diff
@@ -163,7 +163,7 @@ class RequestQuotaTest extends BaseRequestTest {
   def testExemptRequestTime(): Unit = {
     // Exclude `DESCRIBE_QUORUM`, maybe it shouldn't be a cluster action
-    val actions = clusterActions -- clusterActionsWithThrottleForBroker -- RequestQuotaTest.Envelope -- RequestQuotaTest.ShareGroupState - ApiKeys.DESCRIBE_QUORUM
+    val actions = clusterActions -- clusterActionsWithThrottleForBroker -- RequestQuotaTest.Envelope -- RequestQuotaTest.ShareGroupState - ApiKeys.DESCRIBE_QUORUM - ApiKeys.GET_REPLICA_LOG_INFO
```
hm, what determines if we an action/request should be exempt / tracked by the exempt-request-time metric?

ahuang98 replied:
oh interesting, is it related to which apis are throttle-able (in which case we want all new apis to not be exempt from throttling, even cluster actions?)

josefk31 replied:
Yes that is correct, with this one I followed the recent pattern and the KIP-966 specification. The KIP specifies that the request has throttling and recent cluster actions are also throttled. 
I think having throttling does "make sense" in this case - IG its best to avoid asking brokers too many questions about their logs if they are overwhelmed with requests.

ahuang98 commented on core/src/test/scala/unit/kafka/server/RequestQuotaTest.scala:
```diff
@@ -191,7 +191,7 @@ class RequestQuotaTest extends BaseRequestTest {
   }
 
-  private def clusterActionsWithThrottleForBroker: Set[ApiKeys] = {
+  private def clusterActionsWithThrottleForBroker: Set[ApiKeys] = {RequestQuotaTest
```
typo?

ahuang98 commented on core/src/test/scala/unit/kafka/server/KafkaApisTest.scala:
```diff
@@ -13321,6 +13323,284 @@ class KafkaApisTest extends Logging {
     assertEquals(alterShareGroupOffsetsResponseData, response.data)
   }
 
+  def verifyGetReplicaLogInfoRequest(builder: GetReplicaLogInfoRequest.Builder, withResponse: (GetReplicaLogInfoResponse => Unit)): Unit = {
```
could we simplify this? it seems like we could replace `withResponse` with the expected  GetReplicaLogInfoResponseData

ahuang98 commented on core/src/test/scala/unit/kafka/server/KafkaApisTest.scala:
```diff
@@ -13321,6 +13323,284 @@ class KafkaApisTest extends Logging {
+  }
+
+  @Test
+  def testGetReplicaLogInfoFailedToRetrievePartition(): Unit = {
```
nit: this test seems to test the happy case too (getReplicaLogInfo succeeds in retrieving partition) - should we split them out or is the happy case already redundant with other tests?

josefk31 replied:
Good catch, seems redundant so I removed happy case to simplify.

ahuang98 commented on core/src/test/scala/unit/kafka/server/KafkaApisTest.scala:
```diff
@@ -13321,6 +13323,284 @@ class KafkaApisTest extends Logging {
+  }
+
+  @Test
+  def testGetReplicaLogInfoUnknownTopic(): Unit = {
```
may be worth having one of these error-case tests confirm that requests with a mix of "good" and "bad" topics/partitions will not impact the return of information of the "good" partition

ahuang98 commented on core/src/test/scala/unit/kafka/server/KafkaApisTest.scala:
```diff
@@ -13321,6 +13323,284 @@ class KafkaApisTest extends Logging {
+  @Test
+  def testGetReplicaLogInfoRequestTooManyTopics(): Unit = {
+    // 100 topics, 20 partitions per topic = 2k topic-partitions
+    // only first 1000 should be sent back and truncated = true
```
don't forget to change the comment to reflect the new field name

ahuang98 commented on core/src/test/scala/unit/kafka/server/KafkaApisTest.scala:
```diff
@@ -13321,6 +13323,284 @@ class KafkaApisTest extends Logging {
+
+    val expectedResponseData =
+      new GetReplicaLogInfoResponseData().setHasMoreData(true).setBrokerEpoch(brokerEpoch)
+    uuids.take(50).zipWithIndex.foreach { case (uuid, idx) =>
```
nit, could you highlight that we only include the first 50 topics / 1000 partitions with a comment? I know you already have a comment at the top of the test which says a similar thing, but this would help folks find the logic faster

josefk31 committed:
Removed Collections.singletonlist to fix build

ahuang98 commented on core/src/test/scala/unit/kafka/server/KafkaApisTest.scala:
```diff
@@ -13321,6 +13323,284 @@ class KafkaApisTest extends Logging {
+  }
+
+  @Test
+  def testGetReplicaInfoRequestHappyTrail(): Unit = {
```
hehe happy trail, never heard that one before 🥾

josefk31 replied:
Bc all code has a path lol

josefk31 committed:
Updated unit tests in accordance with comments

CalvinLiu7947 commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -261,6 +264,62 @@ class KafkaApis(val requestChannel: RequestChannel,
+      }
+      responseData.topicPartitionLogInfoList().add(topicPartitionLogInfo)
+    }
+    responseData.setHasMoreData(partitionCount == GetReplicaLogInfoRequest.MAX_PARTITIONS_PER_REQUEST)
```
What if the partition count is right at GetReplicaLogInfoRequest.MAX_PARTITIONS_PER_REQUEST? Should we still return HasMoreData = true?

josefk31 replied:
Good catch! Did not think of this edge case.

cmccabe replied:
> What if the partition count is right at GetReplicaLogInfoRequest.MAX_PARTITIONS_PER_REQUEST? Should we still return HasMoreData = true?

no, we shouldn't :)

CalvinLiu7947 reviewed (commented):
LGTM, just one minor comment.
Can you help update the KIP for the API changes? Thanks!

ahuang98 reviewed (approved):
thanks for the changes Jonah!

cmccabe commented on clients/src/main/java/org/apache/kafka/common/protocol/ApiKeys.java:
```diff
@@ -134,8 +134,8 @@ public enum ApiKeys {
-    DELETE_SHARE_GROUP_OFFSETS(ApiMessageType.DELETE_SHARE_GROUP_OFFSETS);
-    
+    DELETE_SHARE_GROUP_OFFSETS(ApiMessageType.DELETE_SHARE_GROUP_OFFSETS),
+    GET_REPLICA_LOG_INFO(ApiMessageType.GET_REPLICA_LOG_INFO, true);
```
Do we have to make this a clusterAction type RPC? It seems like that will disable all throttling, and I don't know that we really need to do that.

josefk31 replied:
This does not need to be a `clusterAction` but keeping it that way makes handling authentication easier to implement as I'd argue we would have to follow the same ACLs model as `fetch` requests. 

It could be something which is added in future IMO. 

It seems like throttling is enabled based on `clientId` which would be whatever broker/controller on cluster is sending these requests which is why I needed to add this [exception](https://github.com/apache/kafka/pull/19664/files#r2114220408).

cmccabe reviewed (approved):
LGTM

josefk31 committed:
GetReplicaLogInfo is no longer a cluster-action

ahuang98 commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -265,33 +265,52 @@ class KafkaApis(val requestChannel: RequestChannel,
+      }
+    }
+
+    val isClusterAction = authorizeClusterOperation(request, CLUSTER_ACTION)
```
nit: `isClusterAction` doesn't seem to describe what this val represents. could this be renamed to something like `isAuthorizedAsClusterAction`? know it's a bit bulky but it's a bit more accurate

josefk31 replied:
Updated!

ahuang98 commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -265,33 +265,52 @@ class KafkaApis(val requestChannel: RequestChannel,
+
+    val isClusterAction = authorizeClusterOperation(request, CLUSTER_ACTION)
+    def isAuthorized(topicName: String): Boolean =
+      isClusterAction || authHelper.authorize(request.context, DESCRIBE, TOPIC, topicName)
```
I would have thought that anyone with CLUSTER_ACTION permission would have DESCRIBE permission (similar to how I would think that ALTER would necessitate DESCRIBE), but guess that kind of ACL dependency is not actually enforced in code.

ahuang98 replied:
I mention this because I wonder if it's extraneous to check `CLUSTER_ACTION` before `DESCRIBE`

josefk31 replied:
This is an interesting comment - I did a bit more research on how `ACLs` work in Kafka. If I understand the comment correctly, there would need to be a system of hierarchy between `CLUSTER_ACTION` and `DESCRIBE` as well as a hierarchy between resources of `CLUSTER` and `TOPIC`.
In other words "Principal P has CLUSTER_ACTION on CLUSTER => Principal P has DESCRIBE on TOPIC"
It seems like `RESOURCES` are not hierarchical. So `TOPIC` and `CLUSTER` are seen as distinct entities in the ACL system. Which makes sense IG? It's like separating the "control-plane" and "data-plane" resources. 

There is a concept of an "implied action" - so if you have `READ` operation then it's implied that you also have `DESCRIBE` permission but I don't think this applies here.

That being said, brokers and controllers can be considered super-users of the cluster; which means that these ACLs kind of "don't matter" for them; they have permission to do everything anyways in the `StandardAuthorizer`. 

I noticed that this is how `handleOffsetForLeaderEpoch` [handles](https://github.com/apache/kafka/blob/18a276edf05ded12e9b7381f6104749ef96f10c8/core/src/main/scala/kafka/server/KafkaApis.scala#L2186) its ACLs - it also checks `CLUSTER_ACTION` and then falls back to `DESCRIBE`.

ahuang98 commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -265,33 +265,52 @@ class KafkaApis(val requestChannel: RequestChannel,
+    while (topicIter.hasNext && partitionCount < GetReplicaLogInfoRequest.MAX_PARTITIONS_PER_REQUEST) {
+      val topic = topicIter.next()
       val partitionIter = topic.partitions().iterator()
+      maybePartitionIter = Some(partitionIter)
```
ah, I was confused why we needed both the partitionIter and the wrapped optional version of it, but I see how you use them differently.

wonder if it might be a bit more readable if we did something like the following instead?
```
maybePartitionIter = Some(topic.partitions().iterator())
val partitionIter = maybePartitionIter.getOrElse(throw new...
```

josefk31 replied:
Yeah; that makes this a lot more clear. Will update the PR.

josefk31 replied:
I ended up renaming it to `previousPartitionIter` which I think points to the variables purpose much more.

ahuang98 reviewed (commented):
few more comments!

josefk31 committed:
Addressed PR comment

CalvinLiu7947 reviewed (approved):
LGTM, thanks!
